e2e/investigation.md

```diff
@@ -0,0 +1,103 @@
+
+## Findings
+
+It was established during the study that the actual problem is that the cleanup and
+re-installation of the development environment (VMs and management cluster).
+
+It would be very time consuming to run the cleanup and the re-installation process after each e2e test.
+
+Because of the nature of the e2e tests there are inherent issues:
+
+    Not every test can run parallel to other tests in the same metal3-dev-env instance e.g.
+    pivoting needs to be executed before re-pivoting. Other example Updating the management cluster or
+    other objects in the cluster during an "upgrade test" might cause stability issues for other
+    tests in the same environment.
+
+    At the moment the cleanup process for e2e tests in CAPM3 is not test specific.
+    There is no cleanup process implemented for individual e2e tests that would restore the dev-env to a state
+    from where other e2e tests could be started, this is partly the case because the restoration process is highly
+    dependent on the e2e test that was executed as some tests affect the configuration of the whole dev-env.
+
+It has become clear that there are multiple ways to achieve the parallel test execution
+
+## Possible solutions
+
+**Option 1 (pool of dev-envs)**:
+
+    This solution requires the bootstrapping of separate dev-env instances
+    (both management cluster and target cluster) for individual e2e test cases.
+    (CON: it scales poorly and it would cause problems with locally running tests on a developer's machine)
+
+**Option 2 (solve it in CI)**:
+
+    This approach would require minimal modification to the CAPM3 Makefile and
+    additional some additional logic to the test code to implement the ability to selectively execute e2e test
+    specs (similarly how it is in dev-env integration tests). Option 3 also includes the individually executable tests.
+
+    This approach would basically be implemented by running multiple e2e test jobs in CI that would only test individual e2e
+    test cases. (CON: this would still require multiple hosts in CityCloud and it would be still slow on local dev-envs).
+
+**Option3 (GO approach)**:
+
+Ginkgo test specs can be executed in a parallel fashion as described [here](https://onsi.github.io/ginkgo/#parallel-specs).
+
+    Each individual test case should have it's own cleanup process that reverts the modifications
+    done by the test execution.
+
+    The cleanup process should aim to avoid re-installation of the
+    dev-env. In some cases like pivoting the cleanup process could be implemented as an other
+    test spec/case e.g "re-pivoting".
+
+    Test specs that are executed in parallel mode should be grouped separately from the ones executed sequentially as
+    not all e2e test cases work well together. The tests that are grouped together as parallelly executable would be
+    executed by Ginkgo's native parallel execution feature. The parallelly executed tests would run at the same time
```

Comment:
>  The parallelly executed tests would run at the same time in different namespaces on the same developer environment (same management cluster and same managed cluster(s)).

It is true that Ginkgo has this feature, but I wonder if we can make it with our e2e tests because our tests involve actions that even change the underlying infrastructure (changing the working cluster in Pivoting, scale in then sacle out in node-reuse...).

Reply:
I have answered it under the next question.